sycl/include/CL/sycl/detail/cstr.hpp

```diff
@@ -0,0 +1,143 @@
+//==- cstr.hpp --- SYCL utilities to manipulate string constants     -------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+///  \file
+///  This file contains meta programming utilities to manipulate string
+///  constants Stored in a type like cstr<char, 'a', 'b'>. This is used for
+///  kernel properties because they must be constants and the API reflects that.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SYCL_DETAIL_CSTR_HPP
+#define SYCL_DETAIL_CSTR_HPP
+
+#include "CL/sycl/detail/pi.h"
+#include <tuple>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+
+namespace sycl {
+
+namespace detail {
+
+/// utility to make any functor's operator() be const even if it originally
+/// wasn't. This is useful to bypass the restriction that kernel operator() must
+/// be const.
+template <typename T> struct const_cheating_wrapper {
+  mutable std::decay_t<T> t;
+  template <typename... Ts> auto operator()(Ts &&... ts) const & {
+    return std::forward<T &>(t)(std::forward<Ts>(ts)...);
+  }
+  template <typename... Ts> auto operator()(Ts &&... ts) const && {
+    return std::forward<T &&>(t)(std::forward<Ts>(ts)...);
+  }
+};
+template <typename T> auto make_const_cheater(T &&t) {
+  return const_cheating_wrapper<T>{std::forward<T>(t)};
+}
+
+/// transform a type like: cstr<char, 'a', 'b'>
+/// into a string "ab"
+template <typename> struct StrPacker {};
+
+/// utility type to manipulate string in constant context
+template <typename CharT, CharT... charpack> struct cstr {
+  using Char = CharT;
+  static constexpr unsigned size = sizeof...(charpack);
+  template <unsigned Idx>
+  using at = __type_pack_element<Idx, cstr<CharT, charpack>...>;
```

Comment:
You cannot write such a line uncommented! :-(
You could explain what it does and add some links to the GCC & Clang extensions.

Reply:
i would have considered the names to be clear enough